src/dmqproto/client/request/internal/Consume.d

```diff
@@ -12,6 +12,711 @@
+            (error status)
+
+    ***************************************************************************/
+
+    private bool handleStatusCode ( ubyte status )
+    {
+        auto consume_status = cast(RequestStatusCode)status;
+
+        if ( Consume.handleGlobalStatusCodes(consume_status,
+            this.context, this.conn.remote_address) )
+        {
+            return false; // Global code, e.g. request/version not supported
+        }
+
+        // Consume-specific codes
+        switch (consume_status)
+        {
+            case consume_status.Started:
+                // Expected "request started" code
+                return true;
+
+            // Treat unknown codes as internal errors.
+            case consume_status.Error:
+            default:
+                // The node returned an error code. Notify the user and
+                // end the request.
+                Consume.Notification n;
+                n.node_error = NodeInfo(conn.remote_address);
+                Consume.notify(this.context.user_params, n);
+                return false;
+        }
+
+        assert(false);
+    }
+
+    /***************************************************************************
+
+        Handler policy, called from BatchRequest template to run the request's
+        main handling logic.
+
+    ***************************************************************************/
+
+    private void handle ( )
+    {
+        scope record_stream = this.new RecordStream;
+        scope reader = this.new Reader(record_stream);
+        scope controller = this.new Controller(record_stream);
+
+        this.request_event_dispatcher.eventLoop(this.conn);
+    }
+
+    /***************************************************************************
+
+        Sends a `Stop` message to the node.
+
+        Params:
+            fiber = the fiber to suspend while sending the `Stop` message
+
+    ***************************************************************************/
+
+    private void sendStopMessage ( MessageFiber fiber )
+    {
+        this.request_event_dispatcher.send(
+            fiber,
+            (conn.Payload payload)
+            {
+                payload.addConstant(MessageType_v3.Stop);
+            }
+        );
+    }
+
+    /***************************************************************************
+
+        The fiber that waits for a batch of records to arrive and passes it to
+        the user, then sends the `Continue` message to the node, in a loop.
+        Handles suspending the request through the controller, for resuming call
+        `resume`. Calling `stop` makes this routine terminate after all
+        remaining records have been passed to the user.
+
+    ***************************************************************************/
+
+    private class RecordStream
+    {
+        /// The acquired buffer to store a batch of records.
+        private Const!(void)[]* batch_buffer;
```

Comment:
Would probably be nice to encapsulate `batch_buffer` and `remaining_batch` in a struct, along with methods to push and pop.

Reply:
... and with compression. That would be a new record batcher ;)